ddl/ddl_api.go

```diff
@@ -5822,15 +5822,20 @@ func (d *ddl) AlterTableAlterPartition(ctx sessionctx.Context, ident ast.Ident,
 	}
 
 	extraCnt := map[placement.PeerRoleType]int{}
-	startKey := hex.EncodeToString(codec.EncodeBytes(nil, tablecodec.GenTablePrefix(partitionID)))
-	endKey := hex.EncodeToString(codec.EncodeBytes(nil, tablecodec.GenTablePrefix(partitionID+1)))
+	startKey := hex.EncodeToString(codec.EncodeBytes(nil, tablecodec.GenTableRecordPrefix(partitionID)))
+	endKey := hex.EncodeToString(codec.EncodeBytes(nil, tablecodec.GenTableRecordPrefix(partitionID+1)))
 	newRules := bundle.Rules[:0]
 	for i, rule := range bundle.Rules {
 		// merge all empty constraints
 		if len(rule.LabelConstraints) == 0 {
 			extraCnt[rule.Role] += rule.Count
 			continue
 		}
+		rule.LabelConstraints = append(rule.LabelConstraints, placement.LabelConstraint{
+			Op:     placement.NotIn,
+			Key:    "engine",
+			Values: []string{"tiflash"},
```

Comment:
Define them as constants and replace other occurrences of "engine" and "tiflash" with the constants.

Reply:
2. Duplication has been handled by label check PR.
1&3. yeah, It seems 3 did not count.